### PR TITLE
Union room chat

### PR DESCRIPTION
**Pull request by PikalaxALT**

### CODE REVIEW

File: include/global.h [hunk 1]
```diff
@@ -312,7 +312,7 @@ struct SecretBaseRecord
     /*0x1A9D*/ u8 gender:1;
     /*0x1A9D*/ u8 sbr_field_1_5:1;
     /*0x1A9D*/ u8 sbr_field_1_6:2;
-    /*0x1A9E*/ u8 trainerName[7]; // TODO: Change PLAYER_NAME_LENGTH to 7
+    /*0x1A9E*/ u8 trainerName[7]; // TODO: Change PLAYER_NAME_LENGTH + 1 to 7
```
Comment (jiangzhengwenjz) [inline include/global.h]: update it and remove comment

File: include/global.h [hunk 2]
```diff
@@ -404,7 +404,7 @@ struct UnkMauvilleOldManStruct
     u8 unk_2D95;
     /*0x2D96*/ u16 mauvilleOldMan_ecArray[6];
     /*0x2DA2*/ u16 mauvilleOldMan_ecArray2[6];
-    /*0x2DAE*/ u8 playerName[PLAYER_NAME_LENGTH + 1];
+    /*0x2DAE*/ u8 playerName[PLAYER_NAME_LENGTH + 1 + 1];
```
Comment (jiangzhengwenjz) [inline include/global.h]: maybe just `+ 2`

File: src/union_room_chat_objects.c [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+    .oam = &sOamData_32x16_2,
+    .anims = sSpriteAnimTable_UnionRoomChatIcons,
+    .affineAnims = gDummySpriteAffineAnimTable,
+    .callback = SpriteCallbackDummy
+};
+
+bool32 UnionRoomChat_TryAllocSpriteWork(void)
+{
+    u32 i;
+    for (i = 0; i < 5; i++)
```
Comment (jiangzhengwenjz) [inline src/union_room_chat_objects.c]: `NELEMS(sSpriteSheets)`

File: src/union_room_chat_objects.c [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+        LoadCompressedSpriteSheet(&sSpriteSheets[i]);
+
+    LoadSpritePalette(&sSpritePalette);
+    sWork = Alloc(0x18);
```
Comment (jiangzhengwenjz) [inline src/union_room_chat_objects.c]: `sizeof(*sWork)`

File: src/union_room_chat_objects.c [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+    if (!sWork)
```
Comment (jiangzhengwenjz) [inline src/union_room_chat_objects.c]: `sWork == NULL`

File: src/union_room_chat_objects.c [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+        return FALSE;
+
+    return TRUE;
+}
+
+void UnionRoomChat_FreeSpriteWork(void)
+{
+    if (sWork)
```
Comment (jiangzhengwenjz) [inline src/union_room_chat_objects.c]: `sWork != NULL`

File: src/union_room_chat_display.c [hunk 7]
```diff
@@ -0,0 +1,1338 @@
...
+    int i;
+    u16 left;
+    u16 top;
+    u8 color[3];
+    u8 str[45];
+    u8 *str2;
+
+    FillWindowPixelBuffer(2, PIXEL_FILL(15));
+    page = GetCurrentKeyboardPage();
+    color[0] = 0;
```
Comment (jiangzhengwenjz) [inline src/union_room_chat_display.c]: `TEXT_COLOR_TRANSPARENT`

File: src/union_room_chat_display.c [hunk 8]
```diff
@@ -0,0 +1,1338 @@
...
+    color[1] = 14;
```
Comment (jiangzhengwenjz) [inline src/union_room_chat_display.c]: `TEXT_DYNAMIC_COLOR_5`

File: src/union_room_chat_display.c [hunk 9]
```diff
@@ -0,0 +1,1338 @@
...
+    color[2] = 13;
```
Comment (jiangzhengwenjz) [inline src/union_room_chat_display.c]: `TEXT_DYNAMIC_COLOR_4`

File: src/union_room_chat_display.c [hunk 10]
```diff
@@ -0,0 +1,1338 @@
...
+{
+    FillWindowPixelRect(1, fillValue, x * 8, 1, width * 8, 14);
+}
+
+static void PrintOnWin1Parameterized(u16 x, u8 *str, u8 bgColor, u8 fgColor, u8 shadowColor)
+{
+    u8 color[3];
+    u8 strbuf[35];
+
+    if (bgColor != 0)
```
Comment (jiangzhengwenjz) [inline src/union_room_chat_display.c]: `TEXT_COLOR_TRANSPARENT`

File: src/union_room_chat_display.c [hunk 11]
```diff
@@ -0,0 +1,1338 @@
...
+static void ClearWin3(void)
+{
+    ClearStdWindowAndFrameToTransparent(3, FALSE);
+    ClearWindowTilemap(3);
+}
+
+static void PrintTextOnWin0Colorized(u16 row, u8 *str, u8 colorIdx)
+{
+    u8 color[3];
+    color[0] = 1;
```
Comment (jiangzhengwenjz) [inline src/union_room_chat_display.c]: `TEXT_COLOR_WHITE`

File: src/union_room_chat.c [hunk 12]
```diff
@@ -0,0 +1,1453 @@
...
+        if (!var0 && !var1)
+            sWork->routineState = 0;
+        break;
+    }
+}
+
+static void ChatEntryRoutine_Switch(void)
+{
+    s16 input;
+    int shouldSwitchPages;
```
Comment (jiangzhengwenjz) [inline src/union_room_chat.c]: looks like `bool32`

File: include/data_8479668.h [hunk 13]
```diff
@@ -0,0 +1,35 @@
+#ifndef GUARD_DATA_8479668_H
```
[the remark this answers is not in the document]
Comment (PikalaxALT) [reply include/data_8479668.h]: Not until there is a discussion on what to name it

**Pull request merged**